std/format.c3

```diff
@@ -0,0 +1,145 @@
+@require_once "io.c3"
+@require_once "string.c3"
+@require_once "type_traits.c3"
+@require_once "util.c3"
+
+function [IO] write : (io : IO&, string : StringView) -> TypeIf<StringView, IsStreamableTextIO<IO>> = {
+    let buffer : Span<u8> = io:request_buffer(string.length);
+
+    // TODO: remave this assumption
+    // TODO: maybe stream large strings directly
+    runtime_assert(buffer.length == string.length);
+    for index in range(string.length) {
+        buffer.data[index] = string.data[index];
+    }
+    io:consume_length(buffer.length);
+    return buffer;
+}
+
+function [IO] write : (io : IO&, number : isize) -> TypeIf<StringView, IsStreamableTextIO<IO>> = {
+    let buffer : Span<u8> = io:request_buffer(21); // len(INT_64_MIN) == 21
+    if number == 0 {
+        buffer.data[0] = "0"[0];
+        // BUG: WE CAN RETURN VOID??
+        return {&buffer.data, 1};
+    }
+
+    let is_negative : bool = number < 0;
+    let length : isize = 0;
+    while number != 0 {
+        // 0 -> 48 ascii
+        let next_digit: u8 = as_logical(__builtin_narrow<i8>(abs(number % 10) + 48));
+
+        buffer.data[length] = next_digit;
+
+        number /= 10;
+        length += 1;
+    }
+
+    if is_negative {
+        buffer.data[length] = "-"[0];
+        length += 1;
+    }
+
+    let result : StringView = {&buffer.data, length};
+
+    // XXX: its a bit unfair to call it a `StringView` when its mutated in-place
+    result:reverse();
+    io:consume_length(length);
+    return result;
+}
+
+function [IO] write : (io : IO&, number : u32) -> TypeIf<StringView, IsStreamableTextIO<IO>> = {
```

Comment:
Did you change your mind on "stringify"? I thought the flexibility was nice.

Reply:
Yeah, I thought `stringify` was nice too, the problem is it becomes kinda unwieldy unless we want to let it do dynamic allocations.

Let's say we want to implement something like `sprintf`, how do we know what size of buffer to give it? Is it always the caller's responsibility? That seems like a c-style memory safety issue.

Let's say we allow dynamic allocation (so the `stringify` can choose its own buffer), now file/ tty io will become kinda ugly because instead of writing blocks we've started allocating huge buffers and writing the entire thing at once (also now we've got to pass an allocator to each stringify call).

Using streamables instead doesn't require any dynamic allocation for file-backed IO but without any code duplication we can also use it to write a dynamically allocating stringify/ StringStream:

```rust
let io_global_allocator : ...;

function request_buffer : (dynamic_io : StringStream&, buffer_length : isize) -> Span<u8> = {
	dynamic_io.underlying_vector:grow_capacity(buffer_length);
	return dynamic_io.underlying_vector:data();
}

function stringify : (a : MyType) -> StringView = {
	let io : StringStream = ...;
}
```

Reply:
> Let's say we allow dynamic allocation (so the stringify can choose its own buffer), now file/ tty io will become kinda ugly because instead of writing blocks we've started allocating huge buffers and writing the entire thing at once (also now we've got to pass an allocator to each stringify call).

Hmm... Can't we have `stringify` return a `std::string` equivalent (kinda hard without destructors but let's say we have RAII one day)
Then you can do scatter/gather I/O with `writev` and it's not even that bad